### PR TITLE
Parse and dump shared limits fields

### PR DESCRIPTION
**Pull request by dschuff**
Wat and binary support for shared memory limits fields.

### CODE REVIEW

File: test/parse/module/bad-memory-shared-nomax.txt [hunk 1]
```diff
@@ -0,0 +1,6 @@
+;;; ERROR: 1
+(module (memory 1 shared))
+
+(;; STDERR ;;;
+Halp I'm getting the wrong error!
```
Comment (dschuff) [inline test/parse/module/bad-memory-shared-nomax.txt]: I'm getting some kind of weird error inside PeekMatch here (i.e. we are PeekMatch'ing on Nat and reading "shared").

**Comment by dschuff** [comment]
I should add that this is still sort of WIP; I just realized I forgot binary output, and I have a test failure but right now I'm mostly using it to debug Binaryen output.

File: src/token.h [hunk 2]
```diff
@@ -70,6 +70,7 @@ enum class TokenType {
   Table,
   Then,
   Type,
+  Shared,
```
Comment (binji) [inline src/token.h]: Sort alphabetically

File: src/wast-parser.cc [hunk 3]
```diff
@@ -606,7 +606,7 @@ bool WastParser::ParseAlignOpt(uint32_t* out_align) {
   }
 }
 
-Result WastParser::ParseLimits(Limits* out_limits) {
+Result WastParser::ParseLimits(Limits* out_limits, bool shared_allowed) {
```
Comment (binji) [inline src/wast-parser.cc]: Enum would be nicer

File: src/wast-parser.cc [hunk 4]
```diff
@@ -616,6 +616,16 @@ Result WastParser::ParseLimits(Limits* out_limits) {
   } else {
     out_limits->has_max = false;
   }
+  if (PeekMatch(TokenType::Shared)) {
```
Comment (binji) [inline src/wast-parser.cc]: I updated the spec text to show the text format for this:
https://webassembly.github.io/threads/text/types.html#memory-types

File: test/parse/module/bad-memory-shared-nomax.txt (hunk 1 above)
Comment (binji) [reply test/parse/module/bad-memory-shared-nomax.txt]: Hm, weird. Maybe will go away with text format from spec...? :-)

File: src/wast-parser.cc (hunk 3 above)
Comment (dschuff) [reply src/wast-parser.cc]: Done.

File: src/wast-parser.cc (hunk 4 above)
Comment (dschuff) [reply src/wast-parser.cc]: Updated.

File: test/parse/module/bad-memory-shared-nomax.txt (hunk 1 above)
Comment (dschuff) [reply test/parse/module/bad-memory-shared-nomax.txt]: \o/

File: src/wast-parser.cc [hunk 5]
```diff
@@ -1027,7 +1038,7 @@ Result WastParser::ParseTableModuleField(Module* module) {
     CheckImportOrdering(module);
     auto import = MakeUnique<TableImport>(name);
     CHECK_RESULT(ParseInlineImport(import.get()));
-    CHECK_RESULT(ParseLimits(&import->table.elem_limits));
+    CHECK_RESULT(ParseLimits(&import->table.elem_limits, Sharing::Allowed));
```
Comment (binji) [inline src/wast-parser.cc]: NotAllowed

File: src/wat-writer.cc [hunk 6]
```diff
@@ -1024,6 +1024,8 @@ void WatWriter::WriteException(const Exception& except) {
 }
 
 void WatWriter::WriteLimits(const Limits& limits) {
+  if (limits.is_shared)
```
Comment (binji) [inline src/wat-writer.cc]: Should be `if (limits.is_shared) { WriteOpenSpace("shared"); ... WriteCloseSpace(); }`. Naming is not great, sorry... :-)

See `WriteBeginGlobal` for similar thing for mut flag.

File: src/wast-parser.cc [hunk 7]
```diff
@@ -606,16 +606,26 @@ bool WastParser::ParseAlignOpt(uint32_t* out_align) {
   }
 }
 
-Result WastParser::ParseLimits(Limits* out_limits) {
+Result WastParser::ParseLimits(Limits* out_limits, Sharing allowed) {
```
Comment (binji) [inline src/wast-parser.cc]: nit: out param is usually last

File: src/wast-parser.cc [hunk 8]
```diff
@@ -606,16 +606,26 @@ bool WastParser::ParseAlignOpt(uint32_t* out_align) {
...
   WABT_TRACE(ParseLimits);
+  Token sharing_tok;
+  if (PeekMatch(TokenType::Shared)) {
```
Comment (binji) [inline src/wast-parser.cc]: Should be `"(" "shared" min max ")"` or `min max?`

Comment (dschuff) [reply src/wast-parser.cc]: Oh, so an inner set of parens. got it.

File: src/wast-parser.cc (hunk 5 above)
Comment (dschuff) [reply src/wast-parser.cc]: now N/A

File: src/wat-writer.cc (hunk 6 above)
Comment (dschuff) [reply src/wat-writer.cc]: Done

**Pull request merged**